### ISSUE TITLE
Action Text doesn't save attached files

### ISSUE THREAD

**Issue by bricechapuis** [rails/rails#36982]
### Steps to reproduce
Run rails action_text:install

Add has_rich_text to the model
class Message < ApplicationRecord
  has_rich_text :content
end

Create the form
<%# app/views/messages/_form.html.erb %>
<%= form_with(model: message) do |form| %>
  <div class="field">
    <%= form.label :content %>
    <%= form.rich_text_area :content %>
  </div>
<% end %>

### Expected behavior
Drag and drop a file, it should get saved upon submitting. Right ?

### Actual behavior
When using the Attach Files button, or dragging and dropping a file, the attachment isn't saved upon submitting the form.

### System configuration
**Rails version**: 6.0.0

**Ruby version**: 2.5.3

I'm a newb, so I'm surely missing an obvious point here, but I managed to get pasted image to save, but not attached files. Any ideas ?

Thank you for your help :)

**Comment by lmansur** [comment]
Did you set up ActiveStorage correctly? This means running the migration for the necessary tables (which ActionText already copies for you) and configuring the storages.

More details here: https://edgeguides.rubyonrails.org/active_storage_overview.html

**Comment by bricechapuis** [comment]
I ran the migration and configured the storage locally, as instructed in the guide already, to no avail. Active storage does save something, the rich text from action text basic feature, but don't save any file attached to the rich text.

**Comment by lmansur** [comment]
I followed the basic guides and got it to work:

```ruby
# app/views/messages/_form.html.erb

<%= form_with(model: message) do |form| %>
  <%= form.label :content %> <%= form.rich_text_area :content %>
  <div class="actions">
    <%= form.submit %>
  </div>
<% end %>
```
```ruby
# app/controllers/messages_controller.rb

  def create
    @message = Message.new(message_params)

    respond_to do |format|
      if @message.save
        format.html { redirect_to @message, notice: 'Message was successfully created.' }
      else
        format.html { render :new }
      end
    end
  end

  private

  def message_params
    params.require(:message).permit(:content)
  end
```
```ruby
# app/models/message.rb

class Message < ApplicationRecord
  has_rich_text :content
end
```



```sh
rails action_text:install
rails db:migrate
```

After uploading the image using the editor, I checked in the console the presence of the attachment:

```ruby
embeds = Message.find(1).content.embeds
embed = embed.first
embed.class # => ActiveStorage::Attachment(id: integer, name: string, record_type: string, record_id: integer, blob_id: integer, created_at: datetime)
```

Can you share your controller as well?

**Comment by nortonandreev** [comment]
Hey. It works for me when I do it locally, but it doesn't work on production, where I use aws3 for data storage. I have image attribute which uses active storage and that gets saved, but not images from action text. I am not sure what the issue is...

**Comment by aryalsan** [comment]
I think this is the major problem. It happened with me too. I upgraded rails to use ActiveStorage and all the images disappears after 5 minutes in production while it works fine in development. So I think its not wise idea to use ActiveStorage and ActionText. These action requires additional db quires too.

**Comment by matthuhman** [comment]
@bricechapuis have you had any luck getting the attachments to work? I'm dealing with the same issue- pasted images work, but attachments don't. For now I'm just going to hide the "attach" option in the toolbar but I'm definitely interested in a fix. I've followed both install guides to the letter and can't seem to get it to work.

**Comment by colinBranca** [comment]
I would also be interested in a solution. I have the same issue.
My ActiveStorage works fine, but when I create a post with ActionText (and `rich_text_area`) the text is saved but attachments don't. Attachments seem to be lost between the view and the controller...

**Comment by colinBranca** [comment]
I achieved to fix my problem.
I figured out that when you drag&drop a file, the file is directly uploaded to the storage (in my case a S3 storage). 
For me the problem came from here and I had a _js_ error: `Cross-Origin Request Blocked`. Which was corrected by changing the CORS permission on my S3 bucket (see solution [here](https://stackoverflow.com/questions/49385431/rails-5-2-active-storage-direct-upload-failure-cors-header-access-control-allo)).

Hope it will help you !

**Comment by nortonandreev** [comment]
> I achieved to fix my problem.
> I figured out that when you drag&drop a file, the file is directly uploaded to the storage (in my case a S3 storage).
> For me the problem came from here and I had a _js_ error: `Cross-Origin Request Blocked`. Which was corrected by changing the CORS permission on my S3 bucket (see solution [here](https://stackoverflow.com/questions/49385431/rails-5-2-active-storage-direct-upload-failure-cors-header-access-control-allo)).
> 
> Hope it will help you !

Didn't fix it for me.

**Comment by nortonandreev** [comment]
> I achieved to fix my problem.
> I figured out that when you drag&drop a file, the file is directly uploaded to the storage (in my case a S3 storage).
> For me the problem came from here and I had a _js_ error: `Cross-Origin Request Blocked`. Which was corrected by changing the CORS permission on my S3 bucket (see solution [here](https://stackoverflow.com/questions/49385431/rails-5-2-active-storage-direct-upload-failure-cors-header-access-control-allo)).
> 
> Hope it will help you !

Hey. Indeed the problem was related to CORS, I switched to Google Storage and managed to fix it there.

Another issue I have now is that sometimes I am getting error 400 and I have to refresh multiple times to get the image. Dunno why is it such a headache...

**Comment by lucaszmoraes** [comment]
I just solved with by creating a CORs config on my S3 Space (Digital Ocean Space):

Origin: https://mysite.com
Allowed Methods: GET, PUT, DELETE, POST
Allowed Headers: Content-Type, Content-Length, Content-Md5
Access Control Max Age: 3000

@nortonandreev I think this is a other solution that ou can explore IF you browser error is related to been blocked by CORs.

Source: https://github.com/rails/rails/issues/32581

Docs: https://edgeguides.rubyonrails.org/active_storage_overview.html#cross-origin-resource-sharing-cors-configuration

@colinBranca watch out for permiting everyone ("*") on your CORs config.

**Comment by jailalawat** [comment]
ActionText not working in case of remote form.In above example if I make remote: true then action text content data not be passed in the params

# app/views/messages/_form.html.erb

<%= form_with(model: message, remote: true) do |form| %>
  <%= form.label :content %> <%= form.rich_text_area :content %>
  <div class="actions">
    <%= form.submit %>
  </div>
<% end %>

**Comment by rozhok** [comment]
I've faced with issue that ActionText doesn't persist attachment tag inside the actual content.

The blob is saved and uploaded to storage, but it never rendered since `content` field doesn't contain placeholder tags. 

It looks like this:

```
Started POST "/rails/active_storage/direct_uploads" for 127.0.0.1 at 2020-03-01 10:56:33 +0200
Processing by ActiveStorage::DirectUploadsController#create as JSON
  Parameters: {"blob"=>{"filename"=>"logo.png", "content_type"=>"image/png", "byte_size"=>7181, "checksum"=>"/qNVDftfldzwW9XfIA7b6A=="}, "direct_upload"=>{"blob"=>{"filename"=>"logo.png", "content_type"=>"image/png", "byte_size"=>7181, "checksum"=>"/qNVDftfldzwW9XfIA7b6A=="}}}
   (527.0ms)  BEGIN
  ActiveStorage::Blob Create (60.9ms)  INSERT INTO "active_storage_blobs" ("key", "filename", "content_type", "byte_size", "checksum", "created_at") VALUES ($1, $2, $3, $4, $5, $6) RETURNING "id"  [["key", "kek"], ["filename", "logo.png"], ["content_type", "image/png"], ["byte_size", 7181], ["checksum", "/qNVDftfldzwW9XfIA7b6A=="], ["created_at", "2020-03-01 08:56:33.778642"]]
   (54.6ms)  COMMIT
  S3 Storage (0.9ms) Generated URL for file at key: kkk (https://s3.eu-central-1.amazonaws.com/)
Completed 200 OK in 677ms (Views: 0.2ms | ActiveRecord: 642.5ms | Allocations: 6248)


Started POST "/admin/pages" for 127.0.0.1 at 2020-03-01 10:56:46 +0200
Processing by Admin::PagesController#create as HTML
  Parameters: {"authenticity_token"=>"iKh9bT0u/omited==", "page"=>{"title"=>"Правила користування сервісом", "url"=>"rules", "content"=>"<div>Here should go attachment:<br></div>"}, "commit"=>"Додати"}
  User Load (53.7ms)  SELECT "users".* FROM "users" WHERE "users"."id" = $1 LIMIT $2  [["id", 1], ["LIMIT", 1]]
  ↳ app/controllers/application_controller.rb:12:in `current_user'
   (55.5ms)  SELECT COUNT(*) FROM "pets" WHERE "pets"."is_approved" = $1  [["is_approved", false]]
  ↳ app/controllers/admin/admin_controller.rb:12:in `fetch_pets_to_approve'
   (53.6ms)  BEGIN
  ↳ app/controllers/admin/pages_controller.rb:23:in `create'
  Page Exists? (54.4ms)  SELECT 1 AS one FROM "pages" WHERE "pages"."title" = $1 LIMIT $2  [["title", "Правила користування сервісом"], ["LIMIT", 1]]
  ↳ app/controllers/admin/pages_controller.rb:23:in `create'
  Page Exists? (53.8ms)  SELECT 1 AS one FROM "pages" WHERE "pages"."url" = $1 LIMIT $2  [["url", "rules"], ["LIMIT", 1]]
  ↳ app/controllers/admin/pages_controller.rb:23:in `create'
  Page Create (56.1ms)  INSERT INTO "pages" ("url", "title", "created_at", "updated_at") VALUES ($1, $2, $3, $4) RETURNING "id"  [["url", "rules"], ["title", "Правила користування сервісом"], ["created_at", "2020-03-01 08:56:46.379742"], ["updated_at", "2020-03-01 08:56:46.379742"]]
  ↳ app/controllers/admin/pages_controller.rb:23:in `create'
  ActionText::RichText Create (57.0ms)  INSERT INTO "action_text_rich_texts" ("name", "body", "record_type", "record_id", "created_at", "updated_at") VALUES ($1, $2, $3, $4, $5, $6) RETURNING "id"  [["name", "content"], ["body", "<div>Here should go attachment:<br></div>"], ["record_type", "Page"], ["record_id", 2], ["created_at", "2020-03-01 08:56:46.437305"], ["updated_at", "2020-03-01 08:56:46.437305"]]
  ↳ app/controllers/admin/pages_controller.rb:23:in `create'
  ActiveStorage::Attachment Load (54.1ms)  SELECT "active_storage_attachments".* FROM "active_storage_attachments" WHERE "active_storage_attachments"."record_id" = $1 AND "active_storage_attachments"."record_type" = $2 AND "active_storage_attachments"."name" = $3  [["record_id", 12], ["record_type", "ActionText::RichText"], ["name", "embeds"]]
  ↳ app/controllers/admin/pages_controller.rb:23:in `create'
  Page Update (53.7ms)  UPDATE "pages" SET "updated_at" = $1 WHERE "pages"."id" = $2  [["updated_at", "2020-03-01 08:56:46.495789"], ["id", 2]]
  ↳ app/controllers/admin/pages_controller.rb:23:in `create'
   (55.0ms)  COMMIT
  ↳ app/controllers/admin/pages_controller.rb:23:in `create'
Redirected to http://localhost:3080/admin/pages
Completed 302 Found in 616ms (ActiveRecord: 547.0ms | Allocations: 11103)
```

Is it how it supposed to work?

**Comment by rozhok** [comment]
Actually everything was worked:
during image paste actiontext sends request to direct in-flight file upload. In my case CORS rules on S3 bucket wasn't configured and upload silently failed. That's why my images won't "stored". I've discovered this error randomly, by checking JS console. After CORS fix, attachments started to work as expected.

However, I still had issues with images resizing so eventually I've switched to froala.
